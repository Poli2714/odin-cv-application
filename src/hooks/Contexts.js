import { createContext } from 'react';

const IndexContext = createContext({
  activeIndex: null,
  setActiveIndex: () => {},
});

const KeyContext = createContext({
  activeKey: {
    eduActiveKey: null,
    expActiveKey: null,
  },
  setActiveKey: () => {},
});

const EduErrMsgContext = createContext({
  showEduErrMessage: false,
  setShowEduErrMessage: () => {},
});

const ExpErrMsgContext = createContext({
  showExpErrMessage: false,
  setShowExpErrMessage: () => {},
});

const PersonalInfoContext = createContext({
  personalInfo: {
    profilePicture: '',
    fullName: '',
    currentJobTitle: '',
    email: '',
    mobile: '',
    city: '',
  },
  setPersonalInfo: () => {},
});

const SummaryContext = createContext({
  summary: '',
  setSummary: () => {},
});

const ExperienceContext = createContext({
  experiences: {
    id: {
      jobTitle: '',
      employer: '',
      jobStartDate: '',
      jobEndDate: '',
      jobDescription: '',
    },
  },
  setExperiences: () => {},
});

const EducationContext = createContext({
  educations: {
    id: {
      degree: '',
      university: '',
      eduStartDate: '',
      eduEndDate: '',
      eduInfo: '',
    },
  },
  setEducations: () => {},
});

export {
  EducationContext,
  EduErrMsgContext,
  ExperienceContext,
  ExpErrMsgContext,
  IndexContext,
  KeyContext,
  PersonalInfoContext,
  SummaryContext,
};
