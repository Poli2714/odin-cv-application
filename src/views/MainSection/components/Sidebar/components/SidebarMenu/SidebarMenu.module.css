.sidebarMenu {
  border-right: 1px solid var(--clr-bg-silver);
  height: 100%;
  position: absolute;
  width: var(--sidebar-menu-width);
}
