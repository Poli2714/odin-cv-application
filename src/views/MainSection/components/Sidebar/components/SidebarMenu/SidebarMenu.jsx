import { useContext } from 'react';

import SidebarMenuElement from './components/SidebarMenuElement/SidebarMenuElement';
import { IndexContext } from 'src/hooks/Contexts';
import styles from './SidebarMenu.module.css';

const sidebarMenuList = [
  {
    id: 'personalInfo',
    label: 'Personal Information',
  },
  {
    id: 'summary',
    label: 'Professional Summary',
  },
  {
    id: 'exp',
    label: 'Experience',
  },
  {
    id: 'edu',
    label: 'Education',
  },
];

function SidebarMenu() {
  const { activeIndex, setActiveIndex } = useContext(IndexContext);

  const handleClick = index =>
    activeIndex !== index ? setActiveIndex(index) : setActiveIndex(null);

  return (
    <div data-testid="sidebar-menu" className={styles.sidebarMenu}>
      <ul>
        {sidebarMenuList.map((menuItem, i) => (
          <li key={menuItem.id} onClick={() => handleClick(i)}>
            <SidebarMenuElement
              label={menuItem.label}
              isActive={activeIndex === i}
            />
          </li>
        ))}
      </ul>
    </div>
  );
}

export default SidebarMenu;
