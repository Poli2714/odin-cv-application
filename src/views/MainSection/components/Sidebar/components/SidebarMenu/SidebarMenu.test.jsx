import { expect, test } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useState } from 'react';

import SidebarMenu from './SidebarMenu';
import { IndexContext } from 'src/hooks/Contexts';

function MockParentComponent({ index }) {
  const [activeIndex, setActiveIndex] = useState(index);
  return (
    <IndexContext.Provider value={{ activeIndex, setActiveIndex }}>
      <SidebarMenu />
    </IndexContext.Provider>
  );
}

test('renders Sidebar menu with 4 menu items', () => {
  render(<MockParentComponent index={null} />);

  expect(screen.getByTestId('sidebar-menu')).toMatchSnapshot();
  expect(screen.getAllByRole('listitem').length).toEqual(4);
});

test('all sidebar menu items have expand icon as img src on initial render', () => {
  render(<MockParentComponent index={null} />);

  const images = screen.getAllByRole('img');
  images.forEach(img => {
    expect(img.src).toMatch(/\/expand.svg$/i);
    expect(img.alt).toEqual('Plus sign');
  });
});

test('renders correct img alt when pressed on different menu item', async () => {
  const user = userEvent.setup();

  render(<MockParentComponent index={0} />);
  const images = screen.getAllByRole('img');
  images.forEach((img, i) => {
    if (i === 0) {
      expect(img.src).toMatch(/\/collapse.svg$/i);
      expect(img.alt).toEqual('Minus sign');
    } else {
      expect(img.src).toMatch(/\/expand.svg$/i);
      expect(img.alt).toEqual('Plus sign');
    }
  });

  await user.click(screen.getAllByRole('listitem')[2]);

  images.forEach((img, i) => {
    if (i === 2) {
      expect(img.src).toMatch(/\/collapse.svg$/i);
      expect(img.alt).toEqual('Minus sign');
    } else {
      expect(img.src).toMatch(/\/expand.svg$/i);
      expect(img.alt).toEqual('Plus sign');
    }
  });
});
