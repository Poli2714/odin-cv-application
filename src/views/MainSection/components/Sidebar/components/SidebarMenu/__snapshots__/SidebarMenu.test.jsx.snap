// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders Sidebar menu with 4 menu items 1`] = `
<div
  class="_sidebarMenu_2b758e"
  data-testid="sidebar-menu"
>
  <ul>
    <li>
      <div
        class="_menuElement_c36675"
        data-testid="menu-item"
      >
        Personal Information
        <img
          alt="Plus sign"
          class="_icon_c36675"
          height="20"
          src="/src/assets/expand.svg"
          width="20"
        />
      </div>
    </li>
    <li>
      <div
        class="_menuElement_c36675"
        data-testid="menu-item"
      >
        Professional Summary
        <img
          alt="Plus sign"
          class="_icon_c36675"
          height="20"
          src="/src/assets/expand.svg"
          width="20"
        />
      </div>
    </li>
    <li>
      <div
        class="_menuElement_c36675"
        data-testid="menu-item"
      >
        Experience
        <img
          alt="Plus sign"
          class="_icon_c36675"
          height="20"
          src="/src/assets/expand.svg"
          width="20"
        />
      </div>
    </li>
    <li>
      <div
        class="_menuElement_c36675"
        data-testid="menu-item"
      >
        Education
        <img
          alt="Plus sign"
          class="_icon_c36675"
          height="20"
          src="/src/assets/expand.svg"
          width="20"
        />
      </div>
    </li>
  </ul>
</div>
`;
